k/p/gen: Use `fragmented_vector` for  `fetchable_partition_response`

BenPope opened the pull request:
Avoid oversize allocations by converting `fetchable_partition_response` to ``small_fragment_vector``

Fixes #11017

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.2.x
- [x] v23.1.x
- [ ] v22.3.x

## Release Notes

### Improvements

* Kafka: Avoid very large contiguous allocations during fetch.

StephanDollberg commented on src/v/utils/fragmented_vector.h:
```diff
@@ -166,19 +166,6 @@ class fragmented_vector {
-    /**
-     * Assign from a std::vector.
-     */
-    fragmented_vector& operator=(const std::vector<T>& rhs) noexcept {
```
Removing this because it is not needed anymore or for any other reason?

StephanDollberg commented on src/v/cloud_storage/tests/tx_range_manifest_test.cc:
```diff
@@ -41,7 +41,17 @@ static remote_manifest_path
+make_fragmented_vector(std::initializer_list<T> in) {
+    fragmented_vector<T> ret;
+    for (auto& e : in) {
+        ret.push_back(e);
```
Can one move out of a initializer_list?

dotnwat replied:
my hunch would be yes. but being a test helper, im not sure how much it matters?

BenPope replied:
See #13854

StephanDollberg reviewed (commented):
> It's possible that large_fragment_vector is too far in the wrong direction

I think the general concern is that we shouldn't just blindly dump the fragmented_vector everywhere because it will create lots of memory wastage if we only put something like a couple 8 bytes elements into it (and this is even worse for the large_fragment_vector I guess). 

`fetch_response::partition_response` isn't too large but I guess we are expecting many of those for topics with higher partition counts?

BenPope commented:
> `fetch_response::partition_response` isn't too large but I guess we are expecting many of those for topics with higher partition counts?

Yeah, the oversize allocation was in: ` rptest.scale_tests.many_partitions_test.ManyPartitionsTest.test_many_partitions`

StephanDollberg reviewed (dismissed):
Right makes sense, I guess the question is how much we are pessimizing responses with few partitions. 

If there is no new data on a partition do we return both a metadata response and a partition response?

I guess it's fine either way as long as there is only a single one for each fetch request.

BenPope commented:
> The cover page mentions metadata response, but this change affects the produce path, not metadata path, right?

The fetch response is the primary fix, but I reworked some stuff in the generator, so the metadata response builds into a fragmented_vector, instead of a std::vector for metadata response, so a large alloc and copy is avoided there, too.

> So it's also perf/memory sensitive.

Aye.

BenPope commented:
> I guess the question is how much we are pessimizing responses with few partitions.

Honestly, I think `large_fragment_vector` might not be the right choice for either operation. We've discussed changing `fragmented_vector` to have more `std::vector`-like behaviour for the first fragment (it doesn't make much sense for further fragments as a `std::vector` would roughly double anyway).

@travisdowns as the author of #8469, do you have any thoughts about the pessimisation here?

 here? Towards the end of this PR it should be possible to avoid `enable_fragmentation_resistance` by using the regular path override mechanism for types.

> If there is no new data on a partition do we return both a metadata response and a partition response?

They're separate responses for separate requests. Both occur quite often.

michael-redpanda commented:
I noticed this has been sitting stale for a little bit.  I think @BenPope has an outstanding question for @travisdowns irt if `large_fragmented_vector` is the right "tool for the job".

dotnwat reviewed (approved):
lgtm

dotnwat commented:
not sure if there are more outstanding questions. maybe ping travis?

dotnwat commented:
>@BenPope - yeah, the part where fragmented vector creates a large first segment as soon as the first entry is added is a problem.
> Arguably it's worse here for fetch because this is a much hotter path in general, and it's also involved in the "poll" loop where we wake up periodically to do the whole fetch execution again but then go to sleep if we don't get enough bytes.

Is this just a matter of using a more modest segment size in the fragmented vector? it sounds like between the "poll" case and large fetches the variability is large enough that we need something adaptive? maybe we could ditch the vector entirely and use a boost intrusive list...

BenPope commented:
I've rebased everything but the last commit into https://github.com/redpanda-data/redpanda/pull/13854 with some minor changes.

BenPope commented:
> I've rebased everything but the last commit into #13854 with some minor changes.

I also accidentally pushed a rebase of this branch, so I've rebased again on top of #13854

BenPope commented:
> > Is this just a matter of using a more modest segment size in the fragmented vector? maybe not, as it sounds like, between the "poll" case and large fetches the variability is large enough that we need something adaptive? maybe we could ditch the vector entirely and use a boost intrusive list...
> 
> Right, I think it's exactly a case of using a small segment size, like say 1K. 1K isn't going to bother anyone and since these are small objects we can still fit plenty into a segment so I don't see much of a downside here. Yes there will be more segments, but none of the operations are really O(number_of_segments) anyway (arguably object destruction can be for primitives which avoid O(n) deletion cost otherwise, but oh well).

Done

dotnwat reviewed (approved):
lgtm

piyushredpanda commented:
Known failure: https://github.com/redpanda-data/redpanda/issues/14053

vbotbuildovich commented:
/backport v23.2.x

vbotbuildovich commented:
/backport v23.1.x

vbotbuildovich commented:
Failed to create a backport PR to v23.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-11181-v23.1.x-362 remotes/upstream/v23.1.x
git cherry-pick -x 793d7af15c408ef2ee99cc0c48a95903c3fca19c f33973cd55f9ff1af86dd1e5f484ebfd958ff9b1
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/6488747224).

BenPope commented:
/backport v23.1.x
